File: binding/src/libqmimedata.cpp
#include <QByteArray>
#include <QMetaType>
#include <QMimeData>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QVariant>
#include <qmimedata.h>
#include "libqmimedata.h"
#include "libqmimedata.hxx"

QMimeData* QMimeData_new() {
	 return new VirtualQMimeData();
}

libqt_string QMimeData_Tr(const char* s) {
	QString _ret = QMimeData::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QMimeData_Urls(const QMimeData* self) {
	auto _ret = self->urls();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

void QMimeData_SetUrls(QMimeData* self, const libqt_list urls) {
	self->setUrls(QList<QUrl>());
}

bool QMimeData_HasUrls(const QMimeData* self) {
	return self->hasUrls();
}

libqt_string QMimeData_Text(const QMimeData* self) {
	QString _ret = self->text();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QMimeData_SetText(QMimeData* self, const libqt_string text) {
	self->setText(QString::fromUtf8(text.data, text.len));
}

bool QMimeData_HasText(const QMimeData* self) {
	return self->hasText();
}

libqt_string QMimeData_Html(const QMimeData* self) {
	QString _ret = self->html();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QMimeData_SetHtml(QMimeData* self, const libqt_string html) {
	self->setHtml(QString::fromUtf8(html.data, html.len));
}

bool QMimeData_HasHtml(const QMimeData* self) {
	return self->hasHtml();
}

QVariant* QMimeData_ImageData(const QMimeData* self) {
	return new QVariant(self->imageData());
}

void QMimeData_SetImageData(QMimeData* self, const QVariant* image) {
	self->setImageData(*image);
}

bool QMimeData_HasImage(const QMimeData* self) {
	return self->hasImage();
}

QVariant* QMimeData_ColorData(const QMimeData* self) {
	return new QVariant(self->colorData());
}

void QMimeData_SetColorData(QMimeData* self, const QVariant* color) {
	self->setColorData(*color);
}

bool QMimeData_HasColor(const QMimeData* self) {
	return self->hasColor();
}

libqt_string QMimeData_Data(const QMimeData* self, const libqt_string mimetype) {
	QByteArray _qb = self->data(QString::fromUtf8(mimetype.data, mimetype.len));
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QMimeData_SetData(QMimeData* self, const libqt_string mimetype, const libqt_string data) {
	self->setData(QString::fromUtf8(mimetype.data, mimetype.len), QByteArray(data.data, data.len));
}

void QMimeData_RemoveFormat(QMimeData* self, const libqt_string mimetype) {
	self->removeFormat(QString::fromUtf8(mimetype.data, mimetype.len));
}

bool QMimeData_HasFormat(const QMimeData* self, const libqt_string mimetype) {
	return self->hasFormat(QString::fromUtf8(mimetype.data, mimetype.len));
}

libqt_list QMimeData_Formats(const QMimeData* self) {
	auto _ret = self->formats();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

void QMimeData_Clear(QMimeData* self) {
	self->clear();
}

libqt_string QMimeData_Tr2(const char* s, const char* c) {
	QString _ret = QMimeData::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QMimeData_Tr3(const char* s, const char* c, int n) {
	QString _ret = QMimeData::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
bool QMimeData_QBaseHasFormat(const QMimeData* self, const libqt_string mimetype) {
	auto* vqmimedata = dynamic_cast<const VirtualQMimeData*>(self);
	if (vqmimedata && vqmimedata->isVirtualQMimeData) {
vqmimedata->setQMimeData_HasFormat_IsBase(true);
	return vqmimedata->hasFormat(QString::fromUtf8(mimetype.data, mimetype.len));
}
}

// Auxiliary method to allow providing re-implementation
void QMimeData_OnHasFormat(const QMimeData* self, intptr_t slot) {
	auto* vqmimedata = dynamic_cast<const VirtualQMimeData*>(self);
	if (vqmimedata && vqmimedata->isVirtualQMimeData) {
vqmimedata->setQMimeData_HasFormat_Callback(reinterpret_cast<VirtualQMimeData::QMimeData_HasFormat_Callback>(slot));
}
}

// Base class handler implementation
libqt_list QMimeData_QBaseFormats(const QMimeData* self) {
	auto* vqmimedata = dynamic_cast<const VirtualQMimeData*>(self);
	if (vqmimedata && vqmimedata->isVirtualQMimeData) {
vqmimedata->setQMimeData_Formats_IsBase(true);
	auto _ret = vqmimedata->formats();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}
}

// Auxiliary method to allow providing re-implementation
void QMimeData_OnFormats(const QMimeData* self, intptr_t slot) {
	auto* vqmimedata = dynamic_cast<const VirtualQMimeData*>(self);
	if (vqmimedata && vqmimedata->isVirtualQMimeData) {
vqmimedata->setQMimeData_Formats_Callback(reinterpret_cast<VirtualQMimeData::QMimeData_Formats_Callback>(slot));
}
}

// Derived class handler implementation
QVariant* QMimeData_RetrieveData(const QMimeData* self, const libqt_string mimetype, QMetaType* preferredType) {
	auto* vqmimedata = dynamic_cast<const VirtualQMimeData*>(self);
	if (vqmimedata && vqmimedata->isVirtualQMimeData) {
	return new QVariant(vqmimedata->retrieveData(QString::fromUtf8(mimetype.data, mimetype.len), *preferredType));
	} else {
	return new QVariant(self->QMimeData::retrieveData(QString::fromUtf8(mimetype.data, mimetype.len), *preferredType));
}
}

// Base class handler implementation
QVariant* QMimeData_QBaseRetrieveData(const QMimeData* self, const libqt_string mimetype, QMetaType* preferredType) {
	auto* vqmimedata = dynamic_cast<const VirtualQMimeData*>(self);
	if (vqmimedata && vqmimedata->isVirtualQMimeData) {
vqmimedata->setQMimeData_RetrieveData_IsBase(true);
	return new QVariant(vqmimedata->retrieveData(QString::fromUtf8(mimetype.data, mimetype.len), *preferredType));
}
}

// Auxiliary method to allow providing re-implementation
void QMimeData_OnRetrieveData(const QMimeData* self, intptr_t slot) {
	auto* vqmimedata = dynamic_cast<const VirtualQMimeData*>(self);
	if (vqmimedata && vqmimedata->isVirtualQMimeData) {
vqmimedata->setQMimeData_RetrieveData_Callback(reinterpret_cast<VirtualQMimeData::QMimeData_RetrieveData_Callback>(slot));
}
}

void QMimeData_Delete(QMimeData* self) {
    delete self;
}

File: binding/include/libqmimedata.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qmimedata.html
pub const qmimedata = struct {

    /// New constructs a new QMimeData object.
    pub fn New() QtC.QMimeData {
        return qtc.QMimeData_new();
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeData_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimedata.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Urls(self: ?*anyopaque, ) []const u8 {
        return qtc.QMimeData_Urls(@ptrCast(self));
    }

    pub fn SetUrls(self: ?*anyopaque, urls: []const u8) void {
        qtc.QMimeData_SetUrls(@ptrCast(self), @ptrCast(urls));
    }

    pub fn HasUrls(self: ?*anyopaque, ) bool {
        return qtc.QMimeData_HasUrls(@ptrCast(self));
    }

    pub fn Text(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeData_Text(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimedata.Text: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetText(self: ?*anyopaque, text: []const u8) void {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
qtc.QMimeData_SetText(@ptrCast(self), text_str);
    }

    pub fn HasText(self: ?*anyopaque, ) bool {
        return qtc.QMimeData_HasText(@ptrCast(self));
    }

    pub fn Html(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeData_Html(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimedata.Html: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetHtml(self: ?*anyopaque, html: []const u8) void {
        const html_str = qtc.libqt_string{
    .len = html.len,
    .data = html.ptr,
};
qtc.QMimeData_SetHtml(@ptrCast(self), html_str);
    }

    pub fn HasHtml(self: ?*anyopaque, ) bool {
        return qtc.QMimeData_HasHtml(@ptrCast(self));
    }

    pub fn ImageData(self: ?*anyopaque, ) QtC.QVariant {
        return qtc.QMimeData_ImageData(@ptrCast(self));
    }

    pub fn SetImageData(self: ?*anyopaque, image: ?*anyopaque) void {
        qtc.QMimeData_SetImageData(@ptrCast(self), @ptrCast(image));
    }

    pub fn HasImage(self: ?*anyopaque, ) bool {
        return qtc.QMimeData_HasImage(@ptrCast(self));
    }

    pub fn ColorData(self: ?*anyopaque, ) QtC.QVariant {
        return qtc.QMimeData_ColorData(@ptrCast(self));
    }

    pub fn SetColorData(self: ?*anyopaque, color: ?*anyopaque) void {
        qtc.QMimeData_SetColorData(@ptrCast(self), @ptrCast(color));
    }

    pub fn HasColor(self: ?*anyopaque, ) bool {
        return qtc.QMimeData_HasColor(@ptrCast(self));
    }

    pub fn Data(self: ?*anyopaque, mimetype: []const u8, allocator: std.mem.Allocator) []u8 {
        const mimetype_str = qtc.libqt_string{
    .len = mimetype.len,
    .data = mimetype.ptr,
};
const _bytearray: qtc.libqt_string = qtc.QMimeData_Data(@ptrCast(self), mimetype_str);
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qmimedata.Data: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn SetData(self: ?*anyopaque, mimetype: []const u8, data: []u8) void {
        const mimetype_str = qtc.libqt_string{
    .len = mimetype.len,
    .data = mimetype.ptr,
};
const data_str = qtc.libqt_string{
    .len = data.len,
    .data = data.ptr,
};
qtc.QMimeData_SetData(@ptrCast(self), mimetype_str, data_str);
    }

    pub fn RemoveFormat(self: ?*anyopaque, mimetype: []const u8) void {
        const mimetype_str = qtc.libqt_string{
    .len = mimetype.len,
    .data = mimetype.ptr,
};
qtc.QMimeData_RemoveFormat(@ptrCast(self), mimetype_str);
    }

    pub fn HasFormat(self: ?*anyopaque, mimetype: []const u8) bool {
        const mimetype_str = qtc.libqt_string{
    .len = mimetype.len,
    .data = mimetype.ptr,
};
return qtc.QMimeData_HasFormat(@ptrCast(self), mimetype_str);
    }

    /// Allows for overriding the related default method
    pub fn OnHasFormat(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) bool) void {
        qtc.QMimeData_OnHasFormat(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseHasFormat(self: ?*anyopaque, mimetype: []const u8) bool {
        const mimetype_str = qtc.libqt_string{
    .len = mimetype.len,
    .data = mimetype.ptr,
};
return qtc.QMimeData_QBaseHasFormat(@ptrCast(self), mimetype_str);
    }

    pub fn Formats(self: ?*anyopaque, ) []const u8 {
        return qtc.QMimeData_Formats(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnFormats(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) [*:0]const u8) void {
        qtc.QMimeData_OnFormats(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseFormats(self: ?*anyopaque, ) []const u8 {
        return qtc.QMimeData_QBaseFormats(@ptrCast(self));
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QMimeData_Clear(@ptrCast(self));
    }

    pub fn RetrieveData(self: ?*anyopaque, mimetype: []const u8, preferredType: QtC.QMetaType) QtC.QVariant {
        const mimetype_str = qtc.libqt_string{
    .len = mimetype.len,
    .data = mimetype.ptr,
};
return qtc.QMimeData_RetrieveData(@ptrCast(self), mimetype_str, @ptrCast(preferredType));
    }

    /// Allows for overriding the related default method
    pub fn OnRetrieveData(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8, QtC.QMetaType) callconv(.c) QtC.QVariant) void {
        qtc.QMimeData_OnRetrieveData(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseRetrieveData(self: ?*anyopaque, mimetype: []const u8, preferredType: QtC.QMetaType) QtC.QVariant {
        const mimetype_str = qtc.libqt_string{
    .len = mimetype.len,
    .data = mimetype.ptr,
};
return qtc.QMimeData_QBaseRetrieveData(@ptrCast(self), mimetype_str, @ptrCast(preferredType));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeData_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimedata.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeData_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimedata.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QMimeData_Delete(@ptrCast(self));
    }
};
